test/uri/uri_path_test.cc

```diff
@@ -1,64 +1,276 @@
+  EXPECT_EQ(path, "");
+}
+
+TEST(URI_path_setter, set_path_with_fragment) {
+  sourcemeta::jsontoolkit::URI uri{"https://example.com"};
+
+  uri.path("/foo%20bar#fragment");
+  EXPECT_EQ(uri.path().value(), "/foo%20bar");
+  EXPECT_EQ(uri.recompose(), "https://example.com/foo%20bar");
+
+  std::string path{"/fooz%20bar#fragment"};
+  uri.path(std::move(path));
+  EXPECT_EQ(uri.path().value(), "/fooz%20bar");
+  EXPECT_EQ(uri.recompose(), "https://example.com/fooz%20bar");
+  EXPECT_EQ(path, "");
+}
+
+TEST(URI_path_setter, set_path_with_query_and_fragment) {
+  sourcemeta::jsontoolkit::URI uri{
+      "https://example.com/old?query=value#fragment"};
+
+  uri.path("/new?query=value#fragment");
+  EXPECT_EQ(uri.path().value(), "/new");
+  EXPECT_EQ(uri.recompose(), "https://example.com/new?query=value#fragment");
+
+  std::string path{"/newer?query=value#fragment"};
+  uri.path(std::move(path));
+  EXPECT_EQ(uri.path().value(), "/newer");
+  EXPECT_EQ(uri.recompose(), "https://example.com/newer?query=value#fragment");
+  EXPECT_EQ(path, "");
+}
+
+TEST(URI_path_setter_no_scheme, set_path_on_host_only) {
+  sourcemeta::jsontoolkit::URI uri{"example.com"};
+
+  uri.path("/foo");
+  EXPECT_EQ(uri.path().value(), "foo");
+  EXPECT_EQ(uri.recompose(), "foo");
+
+  std::string path{"/bar"};
+  uri.path(std::move(path));
+  EXPECT_EQ(uri.path().value(), "bar");
+  EXPECT_EQ(uri.recompose(), "bar");
+  EXPECT_EQ(path, "");
+}
+
+TEST(URI_path_setter_no_scheme, replace_existing_path) {
+  sourcemeta::jsontoolkit::URI uri{"example.com/old"};
+
+  uri.path("/new");
+  EXPECT_EQ(uri.path().value(), "new");
+  EXPECT_EQ(uri.recompose(), "new");
+
+  std::string path{"/newer"};
+  uri.path(std::move(path));
+  EXPECT_EQ(uri.path().value(), "newer");
+  EXPECT_EQ(uri.recompose(), "newer");
+  EXPECT_EQ(path, "");
+}
+
+TEST(URI_path_setter_no_scheme, set_empty_path) {
+  sourcemeta::jsontoolkit::URI uri{"example.com/foo"};
+
+  uri.path("");
+  EXPECT_EQ(uri.path().has_value(), false);
+  EXPECT_EQ(uri.recompose(), "");
+
+  std::string empty_path{""};
+  uri.path(std::move(empty_path));
+  EXPECT_EQ(uri.path().has_value(), false);
+  EXPECT_EQ(uri.recompose(), "");
+}
+
+TEST(URI_path_setter_no_scheme, set_path_on_ip_address) {
+  sourcemeta::jsontoolkit::URI uri{"192.168.0.1"};
+
+  uri.path("/admin");
+  EXPECT_EQ(uri.path().value(), "admin");
+  EXPECT_EQ(uri.recompose(), "admin");
+
+  std::string path{"/admin2"};
+  uri.path(std::move(path));
+  EXPECT_EQ(uri.path().value(), "admin2");
+  EXPECT_EQ(uri.recompose(), "admin2");
+  EXPECT_EQ(path, "");
+}
+
+// TODO: dig why scheme return example.com
+// TEST(URI_path_setter_no_scheme, set_path_with_port) {
+//   sourcemeta::jsontoolkit::URI uri{"example.com:8080"};
+//   std::cout << "urn : " << uri.is_urn() << std::endl;
+//   std::cout << "tag : " << uri.is_tag() << std::endl;
+//   std::cout << "scheme : " << uri.scheme().value() << std::endl;
```

Comment:
Sounds like this whole TODO should be on `scheme()` instead? Here we don't really care, right? As `recompose()` will likely do the right thing anyway?

Reply:
I want to test path here:

```
//   uri.path("/test");
//   EXPECT_EQ(uri.path().value(), "test");
```

Reply:
I found a very weird behavior:

```cc
TEST(URI_path_setter_no_scheme, set_path_with_port) {
  sourcemeta::jsontoolkit::URI uri{"example.com:8080"};

  uri.path("/test");
  EXPECT_EQ(uri.path().value(), "/test");
  EXPECT_EQ(uri.recompose(), "example.com:8080/test");
}
```

```
Expected equality of these values:
  uri.recompose()
    Which is: "example.com:///test"
  "example.com:8080/test"

[  FAILED  ] URI_path_setter_no_scheme.set_path_with_port (0 ms)
```

Reply:
Ah, alright. Interesting